idxmgmt/supporter_factory.go

```diff
@@ -56,50 +61,111 @@ func MakeDefaultSupporter(log *logp.Logger, info beat.Info, configRoot *common.C
+	}
 	if configRoot != nil {
 		if err := configRoot.Unpack(&cfg); err != nil {
 			return nil, err
 		}
 	}
 
-	tmplConfig, err := unpackTemplateConfig(cfg.Template)
-	if err != nil {
-		return nil, fmt.Errorf("unpacking template config fails: %+v", err)
+	templateConfig := defaultTemplateConfig()
+	if cfg.Template != nil {
+		if err := cfg.Template.Unpack(&templateConfig); err != nil {
+			return nil, errors.Wrap(err, "unpacking template config failed")
+		}
 	}
 
 	ilmConfig, err := ilm.NewConfig(info, cfg.ILM)
 	if err != nil {
-		return nil, fmt.Errorf("creating ILM config fails: %v", err)
+		return nil, errors.Wrap(err, "creating ILM config fails")
+	}
+
+	var unmanagedIdxCfg unmanaged.Config
+	if cfg.Output.Name() == esKey {
+		if err := cfg.Output.Config().Unpack(&unmanagedIdxCfg); err != nil {
+			return nil, errors.Wrap(err, "failed to unpack output.elasticsearch config")
+		}
+		if err := checkTemplateESSettings(templateConfig, &unmanagedIdxCfg); err != nil {
+			return nil, err
+		}
 	}
 
 	return &IndexManagementConfig{
 		DataStreams: cfg.DataStreams.Enabled(),
-		Template:    tmplConfig,
-		ILM:         ilmConfig,
 		Output:      cfg.Output,
+		Template:    templateConfig,
+		ILM:         ilmConfig,
+
+		unmanagedIdxCfg:                 unmanagedIdxCfg,
+		registerIngestPipelineSpecified: cfg.RegisterIngestPipeline != nil,
+		setupTemplateSpecified:          cfg.Template != nil,
+		ilmSpecified:                    cfg.ILM != nil,
 	}, nil
 }
 
-func unpackTemplateConfig(cfg *common.Config) (template.TemplateConfig, error) {
-	config := template.DefaultConfig()
-	if cfg == nil {
-		return config, nil
+func checkTemplateESSettings(tmplCfg template.TemplateConfig, indexCfg *unmanaged.Config) error {
+	if !tmplCfg.Enabled || indexCfg == nil {
+		return nil
+	}
+	if indexCfg.Index != "" && (tmplCfg.Name == "" || tmplCfg.Pattern == "") {
+		return errors.New("`setup.template.name` and `setup.template.pattern` have to be set if `output.elasticsearch` index name is modified")
+	}
+	return nil
+}
+
+func (cfg *IndexManagementConfig) logWarnings(log *logp.Logger) {
+	if !cfg.DataStreams {
+		return
+	}
+	const format = "`%s` specified, but will be ignored as data streams are enabled"
+	if cfg.setupTemplateSpecified {
+		log.Warnf(format, "setup.template")
+	}
+	if cfg.ilmSpecified {
+		log.Warnf(format, "apm-server.ilm")
+	}
+	if cfg.registerIngestPipelineSpecified {
+		log.Warnf(format, "apm-server.register.ingest.pipeline")
+	}
+	if cfg.unmanagedIdxCfg.Customized() {
+		log.Warnf(format, "output.elasticsearch.{index,indices}")
+	}
+}
+
+// defaultTemplateConfig customises template.DefaultConfig() with index settings.
+// This is only meaningful when data streams are not in use.
+func defaultTemplateConfig() template.TemplateConfig {
+	cfg := template.DefaultConfig()
+	cfg.Settings = template.TemplateSettings{
+		Index: map[string]interface{}{
+			"codec": "best_compression",
+			"mapping.total_fields.limit": 2000,
```

Comment:
```suggestion

			"mapping": map[string]interface{}{
				"total_fields": map[string]interface{}{
					"limit": 2000,
				},
			},
```

Reply:
When running `./apm-server export template` two `mapping` entries would be produced otherwise

Reply:
Ah thanks, I hadn't noticed that. I flattened it because `cfg.Unpack` is panicking when using a map. I'll dig into why that's happening.

Reply:
https://github.com/elastic/go-ucfg/issues/167

Deeper into the rabbit hole we go...